utils/sample_qc.py

```diff
@@ -0,0 +1,487 @@
+                            )
+        )
+    )
+
+    no_y_call_rate_platforms = set()
+    for platform, sites_called in mean_male_y_sites_called.items():
+        if sites_called < min_male_y_sites_called:
+            logger.warn(f"Mean number of sites in males on Y chromosome for platform {platform} is < {min_male_y_sites_called} ({sites_called} sites found). Y call rate filter will NOT be applied for samples on platform {platform}.")
+            no_y_call_rate_platforms.add(platform)
+
+    sex_ht = sex_ht.annotate_globals(y_call_rate_stats=y_call_rate_stats,
+                                     no_y_call_rate_platforms=no_y_call_rate_platforms if no_y_call_rate_platforms else hl.empty_set(hl.tint32))
+    y_female_stats = sex_ht.y_call_rate_stats[(sex_ht.qc_platform, True)]
+    y_male_stats = sex_ht.y_call_rate_stats[(sex_ht.qc_platform, False)]
+    sex_ht = sex_ht.annotate(
+        is_female=(
+            hl.case()
+                .when(sex_ht.no_y_call_rate_platforms.contains(sex_ht.qc_platform), sex_ht.is_female)
+                .when(sex_ht.is_female & ((sex_ht.y_call_rate - y_female_stats.min)/(y_male_stats.max - y_female_stats.min) < max_y_female_call_rate), True)
+                .when(~sex_ht.is_female & ((sex_ht.y_call_rate - y_female_stats.min)/(y_male_stats.max - y_female_stats.min) > min_y_male_call_rate), False)
+                .or_missing()
+        )
+    )
+    sex_ht = sex_ht.annotate_globals(
+        impute_sex_params=hl.struct(
+            male_min_f_stat=male_min_f_stat,
+            female_max_f_stat=female_max_f_stat,
+            min_male_y_sites_called=min_male_y_sites_called,
+            max_y_female_call_rate=max_y_female_call_rate,
+            min_y_male_call_rate=min_y_male_call_rate
+        )
+    )
+
+    sex_ht =  sex_ht.drop('qc_platform')
+    return (sex_ht)
+
+
+def filter_duplicate_samples(
+        relatedness_ht: hl.Table,
+        samples_rankings_ht: hl.Table,
+        rank_ann: str = 'rank'
+):
+    """
+    Creates a HT with duplicated samples sets.
+    Each row is indexed by the sample that is kept and also contains the set of duplicate samples that should be filtered.
+
+    `samples_rankings_ht` is a HT containing a global rank for each of the sample (smaller is better).
+
+    :param Table relatedness_ht: Input relatedness HT
+    :param Table samples_rankings_ht: HT with global rank for each sample
+    :param str rank_ann: Annotation in `samples_ranking_ht` containing each sample global rank (smaller is better).
+    :return: HT with duplicate sample sets, including which to keep/filter
+    :rtype: Table
+    """
+    logger.info("Getting duplicate samples")
+    dups = get_duplicated_samples(relatedness_ht)
+    logger.info(f"Found {len(dups)} duplicate sets.")
+    dups_ht = hl.Table.parallelize([hl.struct(dup_set=i, dups=dups[i]) for i in range(0, len(dups))])
+    dups_ht = dups_ht.explode(dups_ht.dups, name='_dup')
+    dups_ht = dups_ht.key_by(**dups_ht._dup)
```

Comment:
Ha, so this depends on the keys of `relatedness_ht`. Since `pc_relate` uses `Struct` by default, I had assumed that's what would be used, but I'll make it general.

Reply:
This now also supports non-struct types, like `StringExpression`.